### PR TITLE
ADD: reject_by_annotation to NoisyChannels

### LINKED ISSUE #108: Add a "reject_by_annotation" parameter for finding bad channels
I think the functions to find noisy channels could benefit from a `reject_by_annotation` parameter, [as used in MNE-Python](https://mne.tools/dev/auto_tutorials/preprocessing/20_rejecting_bad_data.html#the-reject-by-annotation-parameter).

It would mean that before working on the continuous raw data, `raw.annotations` are checked for any annotations starting with `BAD`, and cutting out these segments from the raw data.

This would be helpful for cases where you have raw data with long block breaks in which the participant moved etc. --> such "break" segments are not diagnostic for how noisy channels are, and could potentially even screw up the overall stats.

WDYT @a-hurst @yjmantilla ?

EDIT: The `reject_by_annotation` code in MNE happens at the [`raw.get_data`](https://mne.tools/stable/generated/mne.io.Raw.html?highlight=get_data#mne.io.Raw.get_data) level, see: https://github.com/mne-tools/mne-python/blob/c3ff05c05c7bfcf765249f27a517803f27f48b4e/mne/io/base.py#L918-L941

### LINKED ISSUE #133: [Feature suggestion] Allow for relevant annotation selection during processing.
Hello,
When using the pipeline, it would be useful for it to allow for relevant annotations or events that should be considered for channel rejection. Typically, the time between trial should be ignored, and I have annotations that encompass relevant time periods. More commonly, researchers use mne's `epochs` to find relevant portions of their data.
Rejecting channels outside of those time periods is irrelevant.

This joins #108. I'm not sure why PREP doesn't support discontinuities, but if that's a limitation that we can circumvent we definitely should.

### PR DESCRIPTION
**Pull request by Randomidous**
# PR Description
Following a discussion [here](https://github.com/sappelhoff/pyprep/issues/108) and [there](https://github.com/sappelhoff/pyprep/issues/133), this PR implements mne's reject_by_annotation to exclude BAD-annotated time segments from channel quality assessment. Bad segments (e.g., participant movement during breaks) are only temporarily excluded from *statistical analysis* while preserving the original continuous data structure in the output. 

Note:
- This has not been tested
- I recommend looking at the changes commit by commit, since there is also a formatting commit changing a lot of things :) 
- closes: https://github.com/sappelhoff/pyprep/issues/108
- closes: https://github.com/sappelhoff/pyprep/issues/133

 

# Merge Checklist

- [ ] the PR has been reviewed and all comments are resolved
- [ ] all [CI][what-is-ci] checks pass
- [ ] (if applicable): the PR description includes the phrase `closes #<issue-number>` to [automatically close an issue][auto-close-documentation]
- [ ] (if applicable): the changes are documented in the changelog [changelog.rst][changelog]
- [ ] (if applicable): new contributors have added themselves to the authors list in the [`CITATION.cff`][citationcff] file

[what-is-ci]: https://help.github.com/en/actions/building-and-testing-code-with-continuous-integration/about-continuous-integration
[auto-close-documentation]: https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
[changelog]: https://github.com/sappelhoff/pyprep/blob/main/docs/changelog.rst
[citationcff]: https://github.com/sappelhoff/pyprep/blob/main/CITATION.cff

### CODE REVIEW

File: docs/matlab_differences.rst [hunk 1]
```diff
@@ -233,6 +233,32 @@ MATLAB PREP, PyPREP will use a Python reimplementation of ``eeg_interp`` instead
...
+segments (e.g., participant movement during breaks) from *statistical analysis*
+while preserving the original continuous data structure in the output.
+
+When ``reject_by_annotation`` is set to ``'omit'``, MNE's
+:meth:`~mne.io.Raw.get_data` is used to concatenate non-BAD segments for
+computing channel quality metrics. The final processed output retains the
+original continuous structure with all time points intact.
+
+This parameter has no equivalent in MATLAB PREP and is not affected by the
+``matlab_strict`` setting.
```
Comment (sappelhoff) [inline docs/matlab_differences.rst]:
```suggestion
This parameter has no equivalent in MATLAB PREP. When ``matlab_strict`` is set to ``True``, ``reject_by_annotation`` is automatically set to ``None``.
```

In matlab strict mode, reject by annotation should not work. So if a user wants reject by annotation AND matlab strict, we should either (1) warn that we are setting reject_by_annotation to None, or (2) error

**Review by sappelhoff — COMMENTED** [review]
nice work @Randomidous !

Few comments apart from the inline one(s).

1. We should explain what we mean by "bad" segments by linking to https://mne.tools/stable/auto_tutorials/preprocessing/20_rejecting_bad_data.html#the-reject-by-annotation-parameter in the docs via something like 

```
See the MNE tutorial on rejecting data spans:
:ref:`mne:tut-reject-data-spans`.
```
(this should work due to interphinx)

And in the docstrings we can be more brief and say that it's any annotation starting with `BAD` or `bad`

2. Thanks for your "caveats" comment. I think we should add a "test" to pyprep that warns a user IF (i) reject_by_annotation is not None AND (ii) the bad segments are "plenty", i.e., not a few "long" bad segments (block breaks), but potentially loads of small bad segments (where somebody tried to filter out e.g., muscle activations via https://mne.tools/stable/generated/mne.preprocessing.annotate_muscle_zscore.html#mne.preprocessing.annotate_muscle_zscore)

**Review by sappelhoff — COMMENTED** [review]
One more point I forgot: I don't think we need the "NaN" option 🤔  in fact, I think it's unnecessarily complicating this. "Omit" should be sufficient.


Oh and of course let's test this a little before merging it. I suggest you run it on your dataset that you are currently working on and compare the performance / sanity WITH and without reject_by_annotation.

**Review by sappelhoff — APPROVED** [review]
Thanks @Randomidous!
